### PR TITLE
Implement GC roots ourselves in Boehm via callback

### PR DESCRIPTION
**Pull request by joncham**
Implement GC roots ourselves in Boehm rather than relying on their concept of roots which is mainly for static data segments in binaries and not implemented on all platforms (cannot remove roots on Windows)

### CODE REVIEW

File: mono/metadata/domain.c [hunk 1]
```diff
@@ -1208,7 +1208,9 @@ mono_domain_free (MonoDomain *domain, gboolean force)
 
 	domain->setup = NULL;
 
+#ifndef HAVE_BOEHM_GC
```
Comment (lambdageek) [inline mono/metadata/domain.c]: Why is this needed?

Comment (lambdageek) [inline mono/metadata/domain.c]: Oh nevermind. I see it matches mono_domain_create

File: mono/metadata/boehm-gc.c [hunk 2]
```diff
@@ -502,11 +514,26 @@ on_gc_heap_resize (size_t new_size)
...
+}
+
 int
 mono_gc_register_root (char *start, size_t size, void *descr, MonoGCRootSource source, const char *msg)
 {
-	/* for some strange reason, they want one extra byte on the end */
-	GC_add_roots (start, start + size + 1);
+	RootData root_data;
+	root_data.start = start;
+	root_data.end = start + size + 1;
```
Comment (lambdageek) [inline mono/metadata/boehm-gc.c]: you lost the comment about the extra byte

Comment (evincarofautumn) [inline mono/metadata/boehm-gc.c]: Should preserve the comment here. I believe the “strange reason”, which isn’t so strange, is that one-past-the-end is a valid pointer to an object in C (and C++), which should cause a value to be conservatively retained.

Comment (joncham) [inline mono/metadata/boehm-gc.c]: Yes, I can add that back.

File: mono/metadata/boehm-gc.c [hunk 3]
```diff
@@ -56,6 +56,14 @@ void *pthread_get_stackaddr_np(pthread_t);
 static gboolean gc_initialized = FALSE;
 static mono_mutex_t mono_gc_lock;
 
+typedef void (*GC_push_other_roots_proc)();
```
Comment (evincarofautumn) [inline mono/metadata/boehm-gc.c]: I suppose this should be `(void)` (nothing), not `()` (anything).

**Comment by joncham** [comment]
Updated with fixes for comments.

**Comment by lambdageek** [comment]
@monojenkins merge

**Comment by lambdageek** [comment]
Failures are unrelated
